### PR TITLE
Prometheus: Improve autocomplete performance and remove disabling of dynamic label lookup

### PR DESCRIPTION
**Pull request by ivanahuckova**
**What this PR does / why we need it**:

This PRs focues on improving Prometheus autocomplete in 3 ways:
 
1. It improves the way we process the labels. Currently we are using `arrays` and `include` method, which can be slow if the labels cardinality is high (too many iteration over the array with label values). 

This PR introduces usage of `sets` instead of arrays which has **MUCH** better performance with high labels cardinality and slightly better performance for low cardinality. (see more profiling in [here](https://docs.google.com/document/d/1CL_0EjvTsmgDf176z5H8f542Qf2d6jvhXJsd5mKa6O4/edit#heading=h.g9hblewhpsy8 ))

2. It removes the dynamic label lookup limit. We used to have limit of 10,000 metrics which determined if label & values lookup was disabled. However, the number of metrics have nothing to do with how many labels and values would be received, if we did a query. As we have significantly improved the processing of labels, number of received labels and values is no longer a problem, as the processing is fast. Moreover, we have recently (~2 months ago) added the sending of time range in queries, which limits the number of received items even more.  

3. To prevent from running out of memory, we have added the limit of 10,000 items (labels, values, metrics) that can be used for suggestions. If the number is higher and we need to slice the array with suggestions to 10,000, we will then inform the user directly in autocomplete. 

Example with limit of 2, if limit is reached:
image

When limit is not reached:
image

**Which issue(s) this PR fixes**:

Related to https://github.com/grafana/grafana/issues/21893

**Next steps:** 
Improving autocomplete, removing 10,000 metrics limit and limit the number of shown labels and values directly. This is going to be part of separate PR.

### CODE REVIEW

File: public/app/plugins/datasource/prometheus/language_provider.ts [hunk 1]
```diff
@@ -21,7 +22,8 @@ const DEFAULT_KEYS = ['job', 'instance'];
 const EMPTY_SELECTOR = '{}';
 const HISTORY_ITEM_COUNT = 5;
 const HISTORY_COUNT_CUTOFF = 1000 * 60 * 60 * 24; // 24h
-export const DEFAULT_LOOKUP_METRICS_THRESHOLD = 10000; // number of metrics defining an installation that's too big
+// Max number of items - metrics, labels, values - that we display. Prevents running out of memory.
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/prometheus/language_provider.ts:26]:
```suggestion
// Max number of items (metrics, labels, values) that we display as suggestions. Prevents running out of memory.
```

**Review by Elfo404 — APPROVED** [review]
Code looks good! 🚀 
Didn't run it but the numbers look great! Great job!

**Review by davkal — APPROVED** [review]
Looks great!

**Comment by jpdstan** [comment]
This PR is a blocker for us to upgrade Grafana, as we rely heavily on the metric name autocomplete feature for metrics discovery. Could we add a flag such that we follow the old behavior, i.e. 10k limit on label/values but no limit on metric names?

**Comment by geekflyer** [comment]
actually adding to @jpdstan's message. I think if you would allow customizing the SUGGESTIONS_LIMIT via the plugin/datasource config or some environment variable, that'd be ideal for us. We have ~35k metric names and the performance is fine for us, even with the old implementation. So if we can just increase this to ~50k via some configuration I think the new behavior and implementation is preferable.

**Comment by geekflyer** [comment]
@ivanahuckova 50k limit would actually be sufficient for us since we have ~35k metric names from what I can see in prod, however it would be best to let that be configurable regardless since we’re bound to hit 50k sooner than later. 
That said the metrics browser is a seperate UI dialog that is not that convenient to access/use like simply using auto completion imo and auto completion is still limited to 10k metric names. In other words Metrics Browser alone doesn't solve our issue.

**Comment by ivanahuckova** [comment]
I have created a feature request for this https://github.com/grafana/grafana/issues/36390

**Pull request merged**